resources/utils/displayValue.ts

```diff
@@ -0,0 +1,132 @@
+import BigNumber from 'bignumber.js'
+import { isHexString } from 'ethers/lib/utils'
+
+const displayUnitMapping = {
+  million: {
+    lowerBound: BigNumber(1000000),
+    upperBound: BigNumber(1000000000),
+    unitDisplay: 'M'
+  },
+  billion: {
+    lowerBound: BigNumber(1000000000),
+    upperBound: BigNumber(1000000000000),
+    unitDisplay: 'B'
+  },
+  trillion: {
+    lowerBound: BigNumber(1000000000000),
+    upperBound: BigNumber(1000000000000000),
+    unitDisplay: 'T'
+  },
+  quadrillion: {
+    lowerBound: BigNumber(1000000000000000),
+    upperBound: BigNumber(Infinity),
+    unitDisplay: 'Q'
+  }
+}
+const maxDisplayValue = BigNumber(999999999999999999999)
+
+function getDisplay (bn: BigNumber, type: string, decimals: number, displayFullValue?: boolean) {
+  const value = bn.decimalPlaces(decimals, BigNumber.ROUND_FLOOR)
+  if (value.isZero()) {
+    return {
+      approximationSymbol: '<',
+      displayValue: BigNumber(`1e-${decimals}`).toFormat()
+    }
+  }
+
+  if (!displayFullValue) {
+    // shorthand display of large numbers
+    for (const [unitName, { lowerBound, upperBound, unitDisplay }] of Object.entries(displayUnitMapping)) {
+      if (value.isGreaterThanOrEqualTo(lowerBound) && value.isLessThan(upperBound)) {
+        const displayMax = value.isGreaterThan(maxDisplayValue)
+        // maximum display value is hard coded because maxDisplayValue is above the bignumber 15sd limit
```

Comment:
BigNumber doesn't have a digit limit, that's kinda the whole point of it. it's only JS primitive numbers that have this limit

Reply:
Not sure I should change the comment for this, given that the author of the bignumber talks about it in these terms?

https://github.com/MikeMcl/bignumber.js/issues/324#issuecomment-1078117812
https://github.com/MikeMcl/bignumber.js/issues/11

Reply:
these all have to do with using JS number primitives. from the [docs](https://mikemcl.github.io/bignumber.js/): "An error will also be thrown if n is of type _number_ with more than 15 significant digits, as calling [toString](https://mikemcl.github.io/bignumber.js/#toS) or [valueOf](https://mikemcl.github.io/bignumber.js/#valueOf) on these numbers may not result in the intended value."

try it out:
```
> const BigNumber = require('bignumber.js')
> BigNumber.DEBUG = true
true
> a = BigNumber('1.1222211111111111111111111111111')
BigNumber {
  s: 1,
  e: 0,
  c: [ 1, 12222111111111, 11111111111111, 11100000000000 ]
}
> a.toString()
'1.1222211111111111111111111111111'
> b = BigNumber(1.1222211111111111111111111111111)
Uncaught:
Error: [BigNumber Error] Number primitive has more than 15 significant digits: 1.1222211111111111
    at new BigNumber (/home/matt/projects/frame/node_modules/bignumber.js/bignumber.js:330:19)
    at BigNumber (/home/matt/projects/frame/node_modules/bignumber.js/bignumber.js:187:45)
```

Reply:
I mean, I get where the issue comes from - we're debating the semantics here.  Fact is, your example shows a BigNumber error, not a JS primitive error.  "May not result in the intended value" is not a result of a limitation imposed by JS, it's a limitation imposed by BigNumber.js due to the likelihood of unexpected behaviour, no?

Reply:
no, it's a limitation of JS numbers, not of BigNumber, BigNumber just throws an error to warn the user in certain situations. it's not semantics -- since we are not creating new BNs from numbers this large there's no reason to enforce this cap

Reply:
Yes but the BigNumber.js error itself is the limitation in usage.  If I use a massive JS primitive it just behaves weirdly (silently).

Reply:
Not sure what you mean by "no reason to enforce this cap".

Reply:
we're enforcing a max value that BN does not itself enforce. if we're worried about creating BNs from large number primitives we can turn on the debug flag but I don't think we do it anywhere.

Reply:
Ok you've completely lost me with this one.  Maybe run a code suggestion by me and we can see if it passes the tests.

Reply:
also it doesn't behave weirdly, it just may get some of the least significant digits wrong because of the way JS handles floating point numbers. it doesn't even do this in all cases, for example:

```
> h = BigNumber(150000000000000000000000000)
BigNumber { s: 1, e: 26, c: [ 1500000000000 ] }
> h.toFormat()
'150,000,000,000,000,000,000,000,000'
```

since we'd always be rounding numbers this large I don't think this limitation would even come into play. I'd rather have that than showing >999,999 when the actual number may be much different

Reply:
Ok, I'll let you handle this one, I'll fix the others

Reply:
also fwiw this could be a good reminder to try not to use number primitives for any type of arithmetic wherever possible

Reply:
Ok, I see what you mean here now 😅